Client/Frontend/Browser/Search/InitialSearchEngines.swift

```diff
@@ -0,0 +1,161 @@
+// Copyright 2020 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+
+/// This is a list of search engines available to the user at first launch.
+/// For user controlled search engines class look for `SearchEngines.swift`
+class InitialSearchEngines {
+    /// Type of search engine available to the user.
+    enum SearchEngineID: String {
+        case google, bing, duckduckgo, yandex, qwant, startpage, yahoo
+    }
+    
+    struct SearchEngine: Equatable {
+        /// ID of the engine, this is also used to look up the xml file of given search engine if `customId` is not provided.
+        let id: SearchEngineID
+        /// Some search engines have regional variations which correspond to different xml files in `SearchPlugins` folder.
+        /// If you provide this custom id, it will be used instead of `regular` id when accessing the open search xml file.
+        var customId: String?
+        
+        // Only `id` mattera when comparing search engines.
+        // This is to prevent adding more than 2 engines of the same type.
+        static func == (lhs: Self, rhs: Self) -> Bool {
+            lhs.id == rhs.id
+        }
+    }
```

Comment:
This is a bit confusing to me, why do we have both an enum for ID's and custom IDs when only one is used.

We should instead just have `SearchEngine` take 1 ID and create a set of static variables that init with the enum IDs such as:

```
struct SearchEngine: Equatable {
  var id: String
  static let google = SearchEngine(id: "google")
  static let bing = SearchEngine(id: "bing")
  // ...
}
```

This allows you to still create a `SearchEngine` with a custom ID but without requiring to check for that ID in places, as well as let you use dot-syntax when trying to check for the base search engines that used to be `SearchEngineID`. Also should fix up type confusion around `defaultSearchEngine` and `priorityEngine` which both seem like they should be `SearchEngine` objects but instead are `SearchEngineID`s

Reply:
`ID` is used to determine general search engine type, like is it Yandex for both Russian Yandex(ru) and Turkish Yandex(tr), 
my goal was to have this id so we are never able to add 2 of the same engine types, like having both yandex-ru and yandex-tr for one client.

`customId` is like a subcategory, usually used for regional search engines

The idea of my flow was:
1. Load default engines, those which are used by everyone
2. See if we should replace any search engine by locale specific ones
3. If yes, i check for search engine with `id` and replace it with the regional SE

In your example, if we have SE with id `google` added then we want to override it with `google-ru` how do you want to check if this SE already exists in the array and replace it?

Reply:
So for the example of finding `google` and override it with `google-ru` it would be pretty much the same as what you have now:

```swift
private func replaceOrInsert(engine: SearchEngine, with replacementEngine: SearchEngine) {
   guard let engineIndex = engines.firstIndex(of: engine) else {
       engines.append(replacementEngine)
       return
   }

   engines[engineIndex] = replacementEngine
}
...
replaceOrInsert(engine: .yahoo, with: .init(id: "yahoo-jp"))
```

The only difference being you are no longer dealing with having 2 ID's. The reason why I got rid of having 2 ID's was because of how you use it here: https://github.com/brave/brave-ios/pull/2889/files#diff-61f7191853dc8966b402988bb553c0edR262 

The very nature of its usage shows that there is no reason for keeping both (assuming I'm not missing a usage of the base ID somewhere?) 

> my goal was to have this id so we are never able to add 2 of the same engine types, like having both yandex-ru and yandex-tr for one client.

I think this usage will still adhere to this goal. It was still possible in your original code to have two of the same with programmer error since you were using an `enum` + `struct` in an array rather than an `OptionSet` of some kind.

Reply:
`replaceOrInsert(engine: .yahoo, with: .init(id: "yahoo-jp"))`

This would not work with multiple overrides, we can detect `yahoo` <-> `yahoo-jp`, but if a new special case appears we can't find the original `yahoo`

At the moment we have 3 layers of overrides,
1. language code + optional region
2. region only, lang code doesn't matter
3. special cases(for now it's only for priority engines)

at each step we might override the search engine

We can rewrite the code to avoid multiple overrides but this can make the code harder to read

Reply:
Ah I see what you mean. I guess that would mean we would have to keep 2 ID's regardless then.

Reply:
I will try to think if there's a cleaner way, i pushed this PR to run tests as xcode12 crashes on simulator :D